File: linkedin-to-sheets-export/agentic/autogen/tools/googlesheet/export_to_sheets.py
"""Google Sheets export function for LinkedIn search results."""

from typing import Dict, Any, List, Optional
from datetime import datetime
from .client import GoogleSheetsClient


HEADERS = [
    "Profile ID",
    "Profile URL",
    "Full Name",
    "First Name",
    "Last Name",
    "Headline",
    "Current Position",
    "Company",
    "Location",
    "Connection Degree",
    "Is Premium",
    "Is Open To Work",
    "Profile Picture",
    "Extracted At",
]


def export_to_sheets(
    people: List[Dict[str, Any]],
    spreadsheet_id: Optional[str] = None,
    spreadsheet_title: Optional[str] = None,
    sheet_name: str = "LinkedIn People",
) -> Dict[str, Any]:
    """Export LinkedIn search results to Google Sheets using OAuth authentication."""
    try:
        client = GoogleSheetsClient()

        if not spreadsheet_title and not spreadsheet_id:
            now = datetime.now()
            spreadsheet_title = f"LinkedIn People Export - {now.strftime('%Y-%m-%d %H:%M')}"

        is_new_sheet = False

        if not spreadsheet_id:
            result = client.create_spreadsheet(spreadsheet_title, sheet_name, HEADERS)
            spreadsheet_id = result["id"]
            spreadsheet_url = result["url"]
            is_new_sheet = True
        else:
            spreadsheet_url = f"https://docs.google.com/spreadsheets/d/{spreadsheet_id}"

        existing_profile_ids = client.get_existing_profile_ids(
            spreadsheet_id, sheet_name
        )

        timestamp = datetime.now().isoformat()
        rows = []
        for person in people:
            profile_id = person.get("profileId", "")
            if profile_id and profile_id not in existing_profile_ids:
                rows.append(
                    [
                        profile_id,
                        person.get("profileUrl", ""),
                        person.get("fullName", ""),
                        person.get("firstName", ""),
                        person.get("lastName", ""),
                        person.get("headline", ""),
                        person.get("currentPosition", ""),
                        person.get("company", ""),
                        person.get("location", ""),
                        person.get("connectionDegree", ""),
                        str(person.get("isPremium", False)),
                        str(person.get("isOpenToWork", False)),
                        person.get("profilePicture", ""),
                        timestamp,
                    ]
                )

        people_skipped = len(people) - len(rows)
        people_added = 0

        if rows:
            client.append_data(spreadsheet_id, sheet_name, rows)
            people_added = len(rows)

        return {
            "success": True,
            "spreadsheet_id": spreadsheet_id,
            "spreadsheet_url": spreadsheet_url,
            "spreadsheet_title": spreadsheet_title or "Existing Spreadsheet",
            "sheet_name": sheet_name,
            "people_added": people_added,
            "people_skipped": people_skipped,
            "is_new_sheet": is_new_sheet,
            "summary": f"{'Created' if is_new_sheet else 'Updated'} spreadsheet with {people_added} new people{' (' + str(people_skipped) + ' duplicates skipped)' if people_skipped > 0 else ''}",
        }

    except ValueError as e:
        return {"success": False, "error": str(e)}
    except Exception as e:
        return {"success": False, "error": f"Export failed: {str(e)}"}

File: linkedin-to-sheets-export/agentic/langGraph/tools/googlesheet/export-to-sheets.ts
/**
 * Google Sheets export tool for LinkedIn search results
 */

import { tool } from "@langchain/core/tools";
import type { Person } from "../types.js";
import { GoogleSheetsClient } from "./client.js";
import { exportToSheetsSchema, HEADERS } from "./schemas.js";

export const exportToSheetsTool = tool(
  async ({ people, spreadsheetId, spreadsheetTitle, sheetName }) => {
    const client = new GoogleSheetsClient();

    const now = new Date();
    const autoTitle = `LinkedIn People Export - ${now
      .toISOString()
      .replace("T", " ")
      .slice(0, 16)}`;

    const finalSpreadsheetTitle = spreadsheetTitle ?? autoTitle;

    let finalSpreadsheetId = spreadsheetId;
    let spreadsheetUrl: string;
    let isNewSheet = false;

    if (!finalSpreadsheetId) {
      const result = await client.createSpreadsheet(
        finalSpreadsheetTitle,
        sheetName,
        HEADERS
      );
      finalSpreadsheetId = result.spreadsheetId;
      spreadsheetUrl = result.spreadsheetUrl;
      isNewSheet = true;
    } else {
      spreadsheetUrl = `https://docs.google.com/spreadsheets/d/${finalSpreadsheetId}`;
    }

    const existingProfileIds = await client.getExistingProfileIds(
      finalSpreadsheetId,
      sheetName
    );

    const timestamp = new Date().toISOString();
    const rows = people
      .filter((person: Person) => !existingProfileIds.has(person.profileId))
      .map((person: Person) => [
        person.profileId,
        person.profileUrl,
        person.fullName,
        person.firstName ?? "",
        person.lastName ?? "",
        person.headline,
        person.currentPosition ?? "",
        person.company,
        person.location,
        person.connectionDegree ?? "",
        person.isPremium ?? "",
        person.isOpenToWork ?? "",
        person.profilePicture ?? "",
        timestamp,
      ]);

    const peopleSkipped = people.length - rows.length;
    let peopleAdded = 0;

    if (rows.length > 0) {
      await client.appendRows(finalSpreadsheetId, sheetName, rows);
      peopleAdded = rows.length;
    }

    if (!finalSpreadsheetId) {
      throw new Error(
        "Spreadsheet ID is required but was not created or provided"
      );
    }

    return JSON.stringify({
      success: true,
      spreadsheetId: finalSpreadsheetId,
      spreadsheetUrl,
      spreadsheetTitle: finalSpreadsheetTitle,
      sheetName,
      peopleAdded,
      peopleSkipped,
      isNewSheet,
      summary: `${isNewSheet ? "Created" : "Updated"} spreadsheet "${finalSpreadsheetTitle}" with ${peopleAdded} new people${
        peopleSkipped > 0 ? ` (${peopleSkipped} duplicates skipped)` : ""
      }`,
    });
  },
  {
    name: "export-to-sheets",
    description:
      "Export LinkedIn search results to Google Sheets. Automatically creates or updates spreadsheet with duplicate detection by Profile ID.",
    schema: exportToSheetsSchema,
  }
);
